Spigot-Server-Patches/0521-Workaround-for-Client-Lag-Spikes-MC-16225.patch

```diff
@@ -0,0 +1,132 @@
+From: MeFisto94 <MeFisto94@users.noreply.github.com>
+Date: Tue, 12 May 2020 23:02:43 +0200
+Subject: [PATCH] Workaround for Client Lag Spikes (MC-16225) When crossing
+ certain chunk boundaries, the client needlessly calculates light maps for
+ chunk neighbours. In some specific map configurations, these calculations
+ cause a 500ms+ freeze on the Client. This patch basically serves as a
+ workaround by sending light maps to the client, so that it doesn't attempt to
+ calculate them. This mitigates the frametime impact to a minimum (but it's
+ still there).
+
+
+diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
+index 8c1f3290d23795b58a30274c9437dc7dc43fa3a1..33cca62d90bf037df1b0b4e901baedb91400aee7 100644
+--- a/src/main/java/net/minecraft/server/Chunk.java
++++ b/src/main/java/net/minecraft/server/Chunk.java
+@@ -277,7 +277,7 @@ public class Chunk implements IChunkAccess {
+ 
+                     // broadcast
+                     Object[] backingSet = inRange.getBackingSet();
+-                    Packet[] chunkPackets = new Packet[2];
++                    Packet[] chunkPackets = new Packet[10];
+                     for (int index = 0, len = backingSet.length; index < len; ++index) {
+                         Object temp = backingSet[index];
+                         if (!(temp instanceof EntityPlayer)) {
+diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+index 81bb26abc520f49de2e916cf2757ac4c710d02cd..de758f873b4736cdfb617d11fdb0d6f61e8f4a4d 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
++++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+@@ -376,7 +376,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                     player.needsChunkCenterUpdate = false;
+                     player.playerConnection.sendPacket(new PacketPlayOutViewCentre(currPosX, currPosZ));
+                 }
+-                PlayerChunkMap.this.sendChunk(player, new ChunkCoordIntPair(rangeX, rangeZ), new Packet[2], false, true); // unloaded, loaded
++                PlayerChunkMap.this.sendChunk(player, new ChunkCoordIntPair(rangeX, rangeZ), new Packet[10], false, true); // unloaded, loaded
+             },
+             (EntityPlayer player, int rangeX, int rangeZ, int currPosX, int currPosZ, int prevPosX, int prevPosZ,
+              com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> newState) -> {
+@@ -1955,12 +1955,92 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+ 
+     }
+ 
++    private static int getLightMask(final Chunk chunk) {
++        final ChunkSection[] chunkSections = chunk.getSections();
++        int mask = 0;
++
++        for (int i = 0; i < chunkSections.length; ++i) {
++            mask |= (ChunkSection.a(chunkSections[i]) ? 0 : 7) << i; // ChunkSection.isEmpty
```

Comment:
seven?

Reply:
Lightmasks have 18 bits, from the -1 (void) section until the 17th (air) section.
Sections go from 0..16. Now whenever a section is not empty, it can potentially change lighting for the section itself, the section below and the section above, hence the bitmask `111b`, which is `7d`.

Reply:
I see
Instead of commenting what the method is, we use obfhelpers (Like this https://github.com/PaperMC/Paper/blob/master/Spigot-Server-Patches/0004-MC-Utils.patch#L4241)

Reply:
Added obf helper, and added comments made here about the 7 value.